src/units/zone_server.rs

```diff
@@ -155,29 +157,38 @@ impl ZoneServerUnit {
 
         let mut addrs = Vec::new();
 
-        for addr in servers {
-            info!("[{unit_name}]: Binding on {addr:?}");
+        for sock_cfg in servers {
+            info!("[{unit_name}]: Binding on {sock_cfg:?}");
 
-            if let SocketConfig::UDP { addr } | SocketConfig::TCPUDP { addr } = addr {
-                let unit_name: Box<str> = unit_name.into();
-                let svc = svc.clone();
-                tokio::spawn(async move {
-                    if let Err(err) = Self::server(ListenAddr::Udp(addr), svc).await {
-                        error!("[{unit_name}]: {err}");
-                    }
-                });
-                addrs.push(ListenAddr::Udp(addr));
+            if let SocketConfig::UDP { addr } | SocketConfig::TCPUDP { addr } = sock_cfg {
+                let listen_addr = acquire_udp_socket(unit_name, &env_sockets, addr);
+                Self::spawn_udp_server(unit_name, svc.clone(), listen_addr);
+                addrs.push(addr.to_string());
             }
 
-            if let SocketConfig::TCP { addr } | SocketConfig::TCPUDP { addr } = addr {
-                let unit_name: Box<str> = unit_name.into();
-                let svc = svc.clone();
-                tokio::spawn(async move {
-                    if let Err(err) = Self::server(ListenAddr::Tcp(addr), svc).await {
-                        error!("[{unit_name}]: {err}");
-                    }
-                });
-                addrs.push(ListenAddr::Tcp(addr));
+            if let SocketConfig::TCP { addr } | SocketConfig::TCPUDP { addr } = sock_cfg {
+                let listen_addr = acquire_tcp_listener(unit_name, &env_sockets, addr);
+                Self::spawn_tcp_server(unit_name, svc.clone(), listen_addr);
+                addrs.push(addr.to_string());
+            }
+        }
+
+        if unit_name == "PS" {
+            // Also listen on any remaining UDP and TCP sockets provided by
```

Comment:
This isn't safe if the other units are still busy acquiring sockets intended for them as they will be consumed here as left overs, but they are not left over, they're just not consumed yet by other units.